### ISSUE TITLE
[VLM] Image bytes are retained after PIL decoding, causing excessive host memory usage

### ISSUE THREAD

**Issue by leooop-al** [verl-project/verl#7391]
### System Info

- verl entrypoint: `python3 -m verl.trainer.main_ppo`
- verl upstream reference: `main` at `09ac37258ea66b0cb69b2738eec3074ea4e7261c`
- Python: 3.12
- PyTorch: `<paste from python scripts/diagnose.py>`
- Ray: `<paste from python scripts/diagnose.py>`
- vLLM: `<paste from python scripts/diagnose.py>`
- Model engine: Megatron
- Rollout engine: vLLM async/server rollout with AgentLoop
- Algorithm: GRPO
- Cluster: 2 nodes × 8 GPUs
- GPU memory: approximately 96 GiB per GPU
- Model: Qwen 35B-A3B multimodal model
- Actor parallelism:
  - TP=2
  - CP=2
  - PP=1
  - EP=8
- Rollout parallelism: TP=4
- `train_batch_size=256`
- `rollout.n=4`
- Maximum prompt length: 16384
- Maximum response length: 8192
- Each sample contains up to 8 images

#### Summary

When a dataset image entry contains a `bytes` field, `RLHFDataset._build_messages()` creates a PIL object from the bytes but keeps the original bytes in the same dictionary:

```python
if "bytes" in image:
    image["image"] = Image.open(BytesIO(image["bytes"]))

content_list.append({"type": "image", **image})
```

After this operation, the image content can contain both representations:

```python
{
    "type": "image",
    "bytes": compressed_image_bytes,
    "image": PIL.Image.Image(...),
    "path": ...,
}
```

The resulting object is stored in `raw_prompt`.

The original compressed bytes are no longer needed by the model after the image has been decoded, but they remain attached to every trajectory and continue through the AgentLoop and trainer data paths.

#### Minimal dataset structure

A representative row contains messages with `<image>` placeholders and image entries such as:

```python
{
    "prompt": [
        {
            "role": "user",
            "content": "<image><image>Describe the images."
        }
    ],
    "images": [
        {
            "bytes": image_0_bytes,
            "path": "/shared/path/image_0.jpg"
        },
        {
            "bytes": image_1_bytes,
            "path": "/shared/path/image_1.jpg"
        }
    ]
}
```

In the production workload, each sample contains up to 8 images.

The redundant representation can be verified without starting training:

```python
sample = dataset[0]

image_items = [
    item
    for message in sample["raw_prompt"]
    for item in message["content"]
    if isinstance(item, dict) and item.get("type") == "image"
]

assert image_items
assert "bytes" in image_items[0]
assert isinstance(image_items[0].get("image"), PIL.Image.Image)

print(image_items[0].keys())
# Contains both "bytes" and "image".
```

#### Relevant configuration

```text
algorithm.adv_estimator=grpo

data.train_batch_size=256
data.max_prompt_length=16384
data.max_response_length=8192
data.image_key=images
data.dataloader_num_workers=2
data.filter_overlong_prompts_workers=8

actor_rollout_ref.rollout.name=vllm
actor_rollout_ref.rollout.n=4
actor_rollout_ref.rollout.tensor_model_parallel_size=4

actor_rollout_ref.actor.megatron.tensor_model_parallel_size=2
actor_rollout_ref.actor.megatron.context_parallel_size=2
actor_rollout_ref.actor.megatron.pipeline_model_parallel_size=1
actor_rollout_ref.actor.megatron.expert_model_parallel_size=8

trainer.nnodes=2
trainer.n_gpus_per_node=8
```

#### Data flow that retains the binary payload

1. `RLHFDataset._build_messages()` converts `image["bytes"]` to `image["image"]` but does not remove `image["bytes"]`.

2. The resulting messages are stored in `raw_prompt`.

3. GRPO repeats each prompt `rollout.n` times:

```python
gen_batch_output = gen_batch.repeat(
    repeat_times=rollout_n,
    interleave=True,
)
```

`DataProto.repeat()` uses `np.repeat()` for object arrays. This initially repeats references inside one process rather than immediately deep-copying every byte buffer, but the repeated objects are later serialized across independent Ray worker boundaries.

4. `AgentLoopManager` chunks the repeated batch and sends it to multiple Ray `AgentLoopWorker` actors:

```python
worker.generate_sequences.remote(chunk)
```

5. After rollout, the original `raw_prompt` is explicitly attached to each AgentLoop output:

```python
output.extra_fields["raw_prompt"] = kwargs["raw_prompt"]
```

Relevant file:

https://github.com/verl-project/verl/blob/09ac37258ea66b0cb69b2738eec3074ea4e7261c/verl/experimental/agent_loop/agent_loop.py

6. The AgentLoop outputs are gathered and concatenated by the controller.

7. `old_log_prob` converts the entire `DataProto`, including `raw_prompt`, into a TensorDict and dispatches it to actor workers:

```python
batch_td = batch.to_tensordict()
batch_td = left_right_2_no_padding(batch_td)
output = self.actor_rollout_wg.compute_log_prob(batch_td)
```

8. `update_actor` repeats the same full-batch conversion and dispatch:

```python
batch_td = batch.to_tensordict()
batch_td = left_right_2_no_padding(batch_td)
actor_output = self.actor_rollout_wg.update_actor(batch_td)
```

Relevant file:

https://github.com/verl-project/verl/blob/09ac37258ea66b0cb69b2738eec3074ea4e7261c/verl/trainer/ppo/ray_trainer.py

9. The distributed dispatch path splits the TensorDict into DP shards and calls `ray.put()` for the complete shards. Therefore `raw_prompt`, although unused by actor forward, is still serialized and transferred.

Relevant files:

- https://github.com/verl-project/verl/blob/09ac37258ea66b0cb69b2738eec3074ea4e7261c/verl/single_controller/base/decorator.py
- https://github.com/verl-project/verl/blob/09ac37258ea66b0cb69b2738eec3074ea4e7261c/verl/utils/ray_utils.py

#### Memory measurements before removing bytes

At `agent_loop_manager.output`, payload instrumentation reported:

```text
key=raw_prompt

unique_total_gib=12.875
unique_bytes_gib=1.095
unique_pil_est_gib=11.779

logical_total_gib=51.501
logical_bytes_gib=4.378
logical_pil_est_gib=47.118
```

The approximately 4× difference between unique and logical sizes corresponds to `rollout.n=4`.

Before actor dispatch, TaskRunner memory had already reached approximately:

```text
rss_gib=430.96
anon_gib=226.27
shmem_gib=204.13
```

During `parallel_put.serialize`, memory continued growing:

```text
sample=1   rss_gib=445.74
sample=4   rss_gib=504.46
sample=7   rss_gib=573.47
sample=10  rss_gib=639.46
```

This caused severe host-memory pressure and prevented the step from progressing normally.

#### Controlled comparison: path-only images

I generated an equivalent dataset where image entries contained only paths accessible from every Ray node:

```python
{
    "path": "/shared/path/image.jpg"
}
```

No model, batch-size, sequence-length, rollout, or parallelism configuration was changed.

After switching to path-only inputs, payload instrumentation showed:

```text
key=raw_prompt
unique_bytes_gib=0.000
unique_pil_est_gib=0.000
unique_pil_images=0
```

Memory growth was substantially reduced and the rollout proceeded into `old_log_prob`.

The multimodal tensors were still constructed successfully:

```text
stage=agent_loop_worker.output
key=multi_modal_inputs
unique_tensor_gib=20.947 to 23.785 per AgentLoopWorker
```

The complete actor input contained:

```text
stage=old_log_prob.input
key=multi_modal_inputs
unique_tensor_gib=176.958
```

This confirms that the improvement did not come from silently dropping image processing. Images were still decoded, converted into `pixel_values`/multimodal tensors, and delivered to the actor input.

#### About shared memory

This report does not claim that Ray never uses shared memory for multimodal tensors.

With the path-only dataset, the controller showed:

```text
multi_modal_inputs=176.958 GiB

rss_gib=359.94
anon_gib=181.02
shmem_gib=178.37
pss_gib=265.53
uss_gib=179.91
```

The approximately 178 GiB `shmem` value indicates that the multimodal tensor payload does enter shared memory.

However, shared memory is not end-to-end zero-copy in this path, and the controller still materializes the complete multimodal payload. The specific bug reported here is narrower: the original image bytes and PIL/raw-prompt representation are unnecessarily retained and transferred in addition to the required multimodal tensors.

### Expected behavior

At least one of the following should hold:

1. If a shared image path is available, `RLHFDataset` should prefer the path and should not eagerly create a PIL object in the dataloader/controller process.

2. If only `bytes` are available, the implementation should avoid retaining both the original bytes and a PIL object in `raw_prompt`.

3. Large binary image payloads should not be returned to the controller inside `raw_prompt` after `multi_modal_inputs` have already been constructed.

4. `compute_log_prob` and `update_actor` dispatch should select only the fields required by actor forward and should not serialize unused fields such as `raw_prompt`.

5. The behavior should remain compatible with path-only, bytes-only, and PIL-based datasets.

Potential implementation directions include:

- Construct a new image payload instead of mutating and expanding the original dictionary.
- Prefer `path`/URI when both `path` and `bytes` are present.
- Defer bytes decoding to the AgentLoopWorker.
- Remove or replace large binary fields after multimodal preprocessing.
- Avoid attaching `raw_prompt` to rollout output unless it is explicitly required for reward computation or rollout logging.
- Select actor-required TensorDict fields before Ray dispatch.

A simple `image.pop("bytes")` immediately after `Image.open()` may not be sufficient as the only fix because PIL performs lazy loading and can retain the backing `BytesIO` object. The ownership and lifetime of the decoded image should be handled explicitly.

**Comment by leooop-al** [comment]
#7400
